### DISCUSSION TITLE
Welcome to lockable-resources-plugin Discussions!

### DISCUSSION THREAD

**Discussion by mPokornyETM** [jenkinsci/lockable-resources-plugin#342] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by lppedd** [comment]
@mPokornyETM I can't seem to be able to open a new discussion, is that on purpose?

**Reply by mPokornyETM** [reply]
> @mPokornyETM I can't seem to be able to open a new discussion, is that on purpose?

maybe some permissiont, what oyu want to duiscuss I can open it for you.

**Reply by lppedd** [reply]
@mPokornyETM it was a question about what's the best approach for locking multiple stages when the execution is conditional.

Example, I currently have the following (simplified) code:
```
stage('Automation tests (wrapper)') {
  when {
    expression {
      params.AUTOMATION_TESTS
    }
  }
  
  options {
    lock('automation-tests')
  }
  
  stages {
    stage('Automation test 1') {
      ...
    }
     
    stage('Automation test 2') {
      ...
    }
  }
}
```

And I've noticed the plugin attempts to lock even if the entire wrapper stage should be skipped (`params.AUTOMATION_TESTS == false`).

**Reply by lppedd** [reply]
@mPokornyETM found out how to do it. I only needed the `beforeOptions` parameter set to `true`!

**Reply by mPokornyETM** [reply]
Maby you can provide a example here: https://github.com/jenkinsci/lockable-resources-plugin/tree/master/src/doc/examples
And best case one test-case here: https://github.com/jenkinsci/lockable-resources-plugin/blob/master/src/test/java/org/jenkins/plugins/lockableresources/DeclarativePipelineTest.java
I can remember similar discussion (in some issue ir PR).  So ti will helps also for other users.

Thx

**Comment by roded** [comment]
Hello, I can't open a discussion either.

We're using the lockable resources status page to allow manually reserving nodes and resources so that they won't be picked up by automation jobs. We've noticed though that such manual reservations are not persisted after a Jenkins restart. 

Should they be persisted after a restart? Could this be a configuration issue? Or does the plugin no handle such cases?

Thanks!

**Reply by mPokornyETM** [reply]
Hi,

this shal works. The status is stored into .xml config file. That means, it shall be also restored after Jenkins restart.
One question. Do you use JCaC ? In that case, it might happens, thaht the JCaC config wins.

BR

Martin

**Reply by roded** [reply]
Thanks @mPokornyETM .
Our issue was that the status was overridden by the CasC configuration.

**Comment by bushoz** [comment]
Hi! I have such a question:
If you have a lot of Lockable Resources, by default, only 10 are displayed on the Lockable Resources page. You can select 10, 25, 50, 100 and ALL. Is it possible to choose something other than 10 by default?

**Reply by mPokornyETM** [reply]
Your last selection has been stored in the browser cache. That means, when you select 25 and you reload the page it shall be shown 25 lines as well.

**Comment by roded** [comment]
Hi,
I have a handful of build VMs on which various jobs can run. These's jobs are constrained by the amount of memory available to them. 
I'd like to be able to lock a certain amount of memory from a single VM (and only from one VM) without having to specify which VM to lock explicitly.

Let's say every available VM has a total of 8gb of memory.
For example, if the VMs' state is:
* VM1: 2gb free
* VM2: 2gb free
* VM3: 4gb free
* VM4: 2gb free

I'd like to be able to lock the 4gb of memory on VM3 (and only them) by using the label associated with all VM resources.
Is this or something like it possible with the lockable resources plugin? Is there another way to accomplish this?
Thanks!

**Reply by mPokornyETM** [reply]
The only one question is, how you want to handle that, on the VM is onky 2 GB fRAM free. When you now that you can set the labels like
on the node and use the node mirroring option (https://github.com/jenkinsci/lockable-resources-plugin?tab=readme-ov-file#node-mirroring) to mirror the nodes into locakable resources.
In that case you can create our pipeline like:
```
lock(label: 'RAM:4GB', quantity: 1) {
  node(env.LOCKED_RESOURCE0) {
    ... do your things ...
  }
}
```

note you can also use logical operator in the label filter:
https://github.com/jenkinsci/lockable-resources-plugin/blob/master/README.md#jenkins-label-parser-allows-sophisticated-filtering

short answer: yes it is possible, and yes we do it also in my company so.

**Reply by roded** [reply]
@mPokornyETM Thanks for the reply.

I think I should've been clearer: I would like to be able to lock any 4GB from any single VM (but not from multiple VMs). 
If no VMs currently have 4GBs of memory available, I'd like the plugin to wait until 4GBs are available on a single VM.

My first intuition was to create a resource per GB on every VM and then ask the plugin to lock a certain number of GBs but I'm not currently sure how to make sure that the locked GBs are from a single VM.

Hope this makes sense.

**Comment by mPokornyETM** [comment]
Ok, it looks like you have 32GB node and want to use multiple executors (probably docker containers) on the same node. Eauch executor need 4GB.
Hm This will be tricky. But good news, my friend need this for her university project. And we try to find some solution now.

But in your case, you can create resource per executor / container / 4GB-block and lock this one. That might works in case of primitive filter. When you need to allocate 8GB from the same nodes (2 times 4GB) it will be really tricky.

resource-name | labels | description
-----------------|--------|-------------
node01-executor01 | 4GB | 1. executor from node 1
node01-executor01 | 4GB | 2. executor from node 1
node01-executor01 | 4GB | 3. executor from node 1
node01-executor01 | 4GB | 4. executor from node 1
node01-executor02 | 4GB | 1. executor from node 2
node01-executor02 | 4GB | 2. executor from node 2
....
```
lock(label: '4GB', quantity: 1) {
  String nodeName = env.LOCKED_RESOURCE0.substring(0, env.LOCKED_RESOURCE0.indexOf('-'));
  node(env.LOCKED_RESOURCE0) {
    ... do your things ...
  }
}
```

**Reply by roded** [reply]
Thanks @mPokornyETM .
That could work for our basic usage. I'll think about it.
Ideally, we'd like to be able to lock multiple resources from mutually exclusive pools.
Looking forward for the availability of such functionality. 

Many thanks.

**Comment by ZehCastroSEMC** [comment]
> ## 👋 Welcome!
> We’re using Discussions as a place to connect with other members of our community. We hope that you:
> 
> * Ask questions you’re wondering about.
> * Share ideas.
> * Engage with other community members.
> * Welcome others and are open-minded. Remember that this is a community we
>   build together 💪.
> 
> To get started, comment below with an introduction of yourself and tell us about what you do with this community.

Seems like opening new topis is disabled.
Is there a way to get "dynamic" resource names in stage (or pipeline) option blocks?

```
stage('Multiple stages that require locked resource') {
  options {
     lock ("${env.SOME_RUNTIME_THING}")
  }
  stages {
    //work with locked resource
  }
}
```
